pkg/action/util/pod_manager.go

```diff
@@ -0,0 +1,124 @@
+
+	// FIXME: Need also update the entries with value as the old pod
+	// E.g., three actions in a row for the same pod, the third one will get an invalid pod name as the pod has been renamed.
+}
+
+func (p *PodCachedManager) GetPodFromDisplayNameOrUUID(kclient *client.Clientset, displayName, uuid string) (*api.Pod, error) {
+	namespace, name, err := podutil.ParsePodDisplayName(displayName)
+	if err != nil {
+		err = fmt.Errorf("Failed to parse Pod display name %s with uid %s: %v", displayName, uuid, err)
+		glog.Errorf(err.Error())
+		return nil, err
+	}
+
+	pod, err := p.getRunningPod(kclient.CoreV1().Pods(namespace), name, uuid)
+	if err != nil {
+		err = fmt.Errorf("Failed to get Pod by %s/%s: %v", namespace, name, err)
+		glog.Errorf(err.Error())
+		return nil, err
+	}
+
+	return pod, err
+}
+
+func (p *PodCachedManager) getRunningPod(podClient v1.PodInterface, name, uid string) (*api.Pod, error) {
+	// Try the cached name first. If it exists, it means some action was applied so it got renamed.
+	// If try the original name first, it may find the pod that is in terminating process from the previous action
+	// as the action doesn't check if the termination process finishes.
+
+	// Pod is not found with the name. Try using the cached name because actions
+	// might have been applied to the pod.
+	if cachedName, ok := p.getLatestValue(name); ok {
+		glog.V(3).Infof("Using cached name %s for pod %s", cachedName, name)
+		pod, err := podutil.GetPodInPhase(podClient, cachedName.(string), api.PodRunning)
+		if err != nil || pod != nil {
+			return pod, err
+		}
+	}
+
+	pod, err := podutil.GetPodInPhase(podClient, name, api.PodRunning)
+	if err != nil || pod != nil {
+		return pod, err
+	}
+
+	// Pod is not found by name. Try using uid to search for the pod.
+	// Try using the cached uid first.
+	if cachedUid, ok := p.getLatestValue(uid); ok {
+		glog.V(3).Infof("Using cached uid %s for pod %s(%s)", cachedUid, name, uid)
+		pod, err = podutil.GetPodInPhaseByUid(podClient, cachedUid.(string), api.PodRunning)
+		if err != nil || pod != nil {
+			return pod, err
+		}
+	}
+
+	pod, err = podutil.GetPodInPhaseByUid(podClient, uid, api.PodRunning)
+	if err != nil || pod != nil {
+		return pod, err
+	}
+
+	return nil, fmt.Errorf("Failed to find Pod with name %s and uid %s", name, uid)
+}
+
+// Need to find the latest pod as there might be multiple action applied to the pod.
+// E.g., If there are three actions in a row for the same pod, the third one need to get the pod name after the change made by the second action.
+//       action1: pod-foo (old) => pod-foo-c (new)
+//       action2: pod-foo (old) => pod-foo-c (from cache) => pod-foo-c-c (new)
+//       action3: pod-foo (old) => pod-foo-c (from cache) => pod-foo-c-c (from cache again) => pod-foo-c-c-c (new)
+func (p *PodCachedManager) getLatestValue(key string) (interface{}, bool) {
+	val, ok := p.podCache.Get(key)
```

Comment:
This segment can be merged into the loop. Plus, what is this loop for? What do we expect to change during this 20 times continuous retry, what use case does this loop cover?

Reply:
To clarify, the loop is not "retry". It is to find the latest pod after multiple actions being applied.
One example is the case that three actions sent to kubeturbo concurrently and all these actions are on the same pod (pod-foo), e.g., Container vcpu resize, vmem resize, and Pod move.

After vcpu resize finishes, the pod will have name "pod-foo-c". The second action, say vmem resize, will find the cached name "pod-foo-c" by key "pod-foo" as designed. After it's done, it will change the pod to have name "pod-foo-c-c".

The loop is for the third action. The move action will use the key "pod-foo" to get the cached name "pod-foo-c". Without the loop, the action will fail since "pod-foo-c" has been changed to "pod-foo-c-c". The loop is to get the latest pod, which is "pod-foo-c-c" in this case. And then, the move action can be applied to the right pod.